### DISCUSSION TITLE
Add support for custom unit systems

### DISCUSSION THREAD

**Discussion by epenet** [home-assistant/architecture#814] [Entity Models]
Currently, the unit system only support metric and imperial (which in fact is assumed to be US)
However, a lot of countries only use imperial units for some measurements, and metric units for others.

Even countries that use mostly imperial/US units do not use those units all the time.
> Right now, this is already problematic for temperatures with 3D printers, which we now, for example, just convert to F, while even our American friends use C for those.

I believe that the current "metric" versus "imperial" simple selector should be replaced with a complex "regional settings" that allows the users to choose their own defaults for things like number format and default units.

Linked discussions:
- automatic conversion: #809
- support for UK volume units: #813

**Comment by epenet** [comment]
One (easy?) option is to expand on the existing unit system and allow users to select the system for each "object"
- temperature unit (°C / °F)
- length unit (metric / imperial)
- wind speed unit (metric / imperial)
- volume unit (metric / imperial / US)
- mass unit (metric / imperial)
- pressure unit (metric / imperial)
- accumulated_precipitation unit (metric / imperial)

**Reply by TomBrien** [reply]
The UK situation can be even more of a mess to be honest. For example most people I know think of long distances (say between cities) in miles but then there is a generational split in short lengths as metres (say distance to end of the road). I suppose my point here is there will not be a perfect solution, except trying to provide sensible regional default and then perhaps allowing per-entity preferences.

**Reply by epenet** [reply]
per-entity is now available (from 2022.10), and will always override the regional settings so it would still be possible to set the default length unit to metric, and override specific sensors to use "miles"

**Comment by epenet** [comment]
Another option is to have an extensive list of all convertible units and let the user decide if they should be converted or not
- Temperature:
  - °C => do not convert
  - °F => convert to °C
  - K => do not convert
- Length:
  - mm => do not convert
  - cm => do not convert
  - m => do not convert
  - km => do not convert
  - in => convert to mm
  - ft => convert to m
  - yd => convert to m
  - mil => convert to km
- Volume:
  - ...

**Comment by epenet** [comment]
Yet another option is to have a list of device classes, and set the default unit that way:
- SensorDeviceClass.DISTANCE: metric / imperial
- SensorDeviceClass.PRECIPITATION: metric / imperial
- SensorDeviceClass.SPEED: metric / imperial
- SensorDeviceClass.TEMPERATURE: °C / °F
- SensorDeviceClass.VOLUME: metric / imperial / US
